# Dolosa Project
This repository provides a framework to replicate the bioinformatics analyses performed in the comparative genomics study of 80 B. dolosa recovered from a single patient. The bioinformatics pipelines can be found in the directory notebook. Each process is described in a jupyter notebook; these notebooks can be used in a local machine as long as the jupyter environment is installed with R, Python and Bash kernels.

The required software can be found in the directory bin. Unless explicitly mentioned, the proper references for these programs can be found in the READ.md file in the bin directory. Otherwise, the code can be replicated as if the starting point was the main folder (DolosaProject).

## Pipeline Steps

[01 Download Data](https://github.com/juliofdiaz/DolosaProject/blob/master/notebook/1_DownloadData.ipynb)

[02 Removal of Duplicates](https://github.com/juliofdiaz/DolosaProject/blob/master/notebook/2_RemoveDuplicates.ipynb)

[03 De novo assembly](https://github.com/juliofdiaz/DolosaProject/blob/master/notebook/3_deNovoAssembly.ipynb)

[04 Remove Contamination](https://github.com/juliofdiaz/DolosaProject/blob/master/notebook/4_RemoveContamination.ipynb)

[05 Sequencing Quality Control]()

[06 Genome Annotation]()

[07 Reference Mapping]()

[08 SNP Calling and Annotation]()

[09 Indel Calling and Annotation]()

[10 Plotting with Circos]()

[11 Nucleotide Diversity]()

[12 Phylogenetic Analysis]()

[13 Structure Analysis]()

[14 Plotting with Fishplot]()

[15 Splitstree Analysis]()

[16 dN over dS Analysis]()

[17 Plotting Antibiotic Resistance]()

[18 Genome Wide Association Study]()

[19 Genes Mutated at least Twice]()

[20 Orthologous Analysis]()

[21 Depth of Coverage Analysis]()

[22 Plasmid Annotation]()
